crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -111,46 +128,93 @@ impl CoinGecko {
             })?
             .0;
 
-        let price = prices
+        Ok(tokens
+            .iter()
+            .map(|token| {
+                (
+                    **token,
+                    prices
+                        .get(token)
+                        .ok_or(PriceEstimationError::NoLiquidity)
+                        .map(|price| price.eth),
+                )
+            })
+            .collect())
+    }
+
+    pub async fn send_request_price_in_eth(&self, token: Token) -> NativePriceEstimateResult {
+        let prices = self.send_bulk_request_price_in_eth(&[&token]).await?;
+        prices
             .get(&token)
-            .ok_or(PriceEstimationError::NoLiquidity)?;
-        Ok(price.eth)
+            .ok_or(PriceEstimationError::NoLiquidity)?
+            .clone()
+    }
+}
+
+#[async_trait]
+impl NativePriceBatchFetcher for CoinGecko {
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> std::result::Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError> {
+        let mut requested_tokens = tokens.iter().collect::<Vec<_>>();
+        match self.quote_token {
+            QuoteToken::Eth => self.send_bulk_request_price_in_eth(&requested_tokens).await,
+            QuoteToken::Other(native_price_token) => {
```

Comment:
nit: the `Other` code path is quite big. Might be nicer to have it in a separate function to improve readability.